Fix test imports

nickdrozd opened the pull request:
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Trying to run `pytest` locally without `git` package installed raises `ModuleNotFoundError`, even though primer is not run locally. Similar for `pytest_benchmark`.

This MR delays imports from `git` until they are actually used. The `pytest_benchmark` import was only used for typechecking, so it is pushed under `TYPE_CHECKING` flag. (This is yet another benefit to separating typechecking imports from runtime imports.)

nickdrozd commented:
Seems to work okay, but I guess the primer files aren't covered? Is there a way to skip the coverage check for this one?

Pierre-Sassoulas reviewed (changes requested):
I don't think there's anything to fix, this is due to pylint not being installed properly:

https://github.com/pylint-dev/pylint/blob/c0ecd70e6ff9e92ee686671c7e0f152ab643f25e/pyproject.toml#L55

A possible solution would be to move the test utils in another package entirely, the name is reserved https://pypi.org/project/pylint-testutils/.

nickdrozd commented:
Okay, I installed the `testutils` and the Git imports works fine. The Pytest benchmark still isn't installed though, so running `pytest` crashes immediately with an import error. I don't know if that's something that needs to be fixed, but the typechecking import can be moved either way.

I dropped the Git import commit and replaced it with an unrelated change that moves `GetProjectCallable` type to `testutils`.

Pierre-Sassoulas commented:
The typing being all guarded is not consensual along pylint core maintainer (there's an ongoing discussion with Marc about this), let's not change anything before it is resolved. 

Also I'm not sure what the issue is, there's a test requirements file that you need to pip ``install -r`` if you want to launch the tests locally ?

https://github.com/pylint-dev/pylint/blob/c0ecd70e6ff9e92ee686671c7e0f152ab643f25e/requirements_test_min.txt#L7

https://pylint.readthedocs.io/en/stable/development_guide/contributor_guide/tests/install.html#basic-installation

nickdrozd commented:
My understanding is that the concern is with using the type guard project-wide. Is there an objection to using the type guard in each and every case? For this particular import, using the guard provides a clear material benefit: it allows running tests without all the test dependencies installed.

Pierre-Sassoulas commented:
There's a clear benefit to installing the test requirements from the test requirements file: you're not going to run into strange issue because you have an out of date dependency and we won't have to complexify the code to accomodate partially installed tests requirements. 

Also introducing unneeded typing guards MR by MR is not very different from introducing them all at once. The discussion still need to happens. I found the 'this is going to create churn' argument convincing and imo the 'default' at the moment is to not have typing guard unless necessary so if Marc thinks strongly that it's not desirable I would be in favor of not changing anything unless we have strong arguments in favor (like major performance improvment, or it becoming the standard).
